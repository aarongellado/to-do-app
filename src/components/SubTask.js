import { Close } from "@mui/icons-material";
import { Box, Checkbox, IconButton, Typography } from "@mui/material";
import { useDispatch } from "react-redux";
import { deleteSubTask, updateSubTaskStatus } from "../redux/list";

export default function SubTask(props) {
    const dispatch = useDispatch();

    const handleRemoveSubtask = () => {
        let payload = {
            mainTaskId: props.mainTaskId,
            index: props.index
        }
        dispatch(deleteSubTask(payload))
    }

    return (
        <Box
            sx={{
                display: "flex",
                flexDirection: 'row',
                border: "1px solid #bbbbbb",
                borderRadius: 2,
                justifyContent: "space-between",
                alignItems: "center",
                padding: "5px 10px"
            }}
            data-testid={`subTask-${props.index}`}
        >
            <Typography>{props.subtaskName}</Typography>
            <Box>
                <Checkbox
                    checked={props.status}
                    onChange={
                        (e) => {
                            dispatch(updateSubTaskStatus({
                                mainTaskId: props.mainTaskId,
                                index: props.index,
                                checked: e.target.checked
                            }))
                        }} />
                <IconButton
                    onClick={() => handleRemoveSubtask()}
                >
                    <Close />
                </IconButton>
            </Box>
        </Box>
    )
}