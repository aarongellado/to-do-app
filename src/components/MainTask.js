import { Add, Close } from "@mui/icons-material";
import { Box, IconButton, Input, TextField, Typography } from "@mui/material";
import Checkbox from '@mui/material/Checkbox';
import { useState } from "react";
import { useDispatch } from "react-redux";
import { newSubTask, deleteItem, updateStatus } from "../redux/list";
import SubTask from "./SubTask";

export default function MainTask(props) {
    const [subtaskValue, setSubtaskValue] = useState('');
    const dispatch = useDispatch();

    const handleAddSubtask = () => {
        let payload = {
            taskName: props.taskName,
            id: props.id,
            status: props.status,
            subtasks: props.subtasks,
            subtaskName: subtaskValue
        }
        dispatch(newSubTask(payload));
        setSubtaskValue('');
    }

    const handleRemoveTask = () => {
        dispatch(deleteItem(props.id));
    }

    return (
        <Box
            sx={{
                width: .23,
                marginRight: "2%"
            }}
            data-testid={`mainTask-${props.id}`}
        >
            {console.log(props)}
            <Box
                className="main-task-container"
                sx={{
                    border: "1px solid #bbbbbb",
                    borderRadius: 2,
                    maxWidth: 1,
                    marginTop: "20px",
                    display: "flex",
                    flexDirection: "row",
                    justifyContent: "space-between",
                    alignItems: "center",
                    paddingLeft: "10px"
                }}
            >
                <Typography variant="h6">{props.taskName}</Typography>
                <Box>
                    <Checkbox
                        checked={props.status}
                        onChange={(e) => {
                            dispatch(updateStatus({
                                id: props.id,
                                status: e.target.checked
                            }))
                        }}
                    />
                    <IconButton onClick={() => handleRemoveTask()}><Close /></IconButton>
                </Box>
            </Box>
            {props.showSubtasks &&
                <Box
                    className="subtask-container"
                    sx={{
                        maxWidth: .95,
                        marginLeft: "2.5%"
                    }}>
                    {props.subtasks.length > 0 &&
                        props.subtasks.map((el, index) => {
                            return (
                                <SubTask
                                    subtaskName={el.subtaskName}
                                    status={el.status}
                                    index={index}
                                    mainTaskId={props.id}
                                />
                            );
                        })
                    }
                    <Box
                        sx={{
                            border: "1px solid #bbbbbb",
                            borderRadius: 2,
                            justifyContent: "space-between",
                            alignItems: "center",
                            padding: "5px 10px"
                        }}
                    >
                        <TextField
                            sx={{ width: "auto" }}
                            size="small"
                            value={subtaskValue}
                            onInput={(e) => setSubtaskValue(e.target.value)}
                        />
                        <IconButton onClick={() => handleAddSubtask()}><Add /></IconButton>
                    </Box>
                </Box>
            }
        </Box>
    )
}