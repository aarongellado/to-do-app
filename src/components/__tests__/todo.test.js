import { render, screen, cleanup } from '@testing-library/react'
import TestRenderer from 'react-test-renderer';
import MainTask from '../MainTask'
import SubTask from '../SubTask'
import { Provider } from 'react-redux'
import store from '../../redux/store'

test('Should render MainTask component', () => {
    const mainTask = {
        taskName: "walk dog",
        id: 1,
        status: false,
        subtasks: [
            {
                subTaskName: "walk through park",
                status: false
            },
            {
                subTaskName: "get back home",
                status: false
            }
        ],
    }
    render(<Provider store={store}><MainTask showSubtasks={mainTask.showSubtasks} taskName={mainTask.taskName} id={mainTask.id} status={mainTask.status} subtasks={mainTask.subtasks} /></Provider>);
    const mainTaskElement = screen.getByTestId('mainTask-1');
    expect(mainTaskElement).toBeInTheDocument();
    expect(mainTaskElement).toHaveTextContent("walk dog");
});

cleanup();

test('Should render SubTask component', () => {
    const subTask = {
        mainTaskId: 1,
        index: 1, 
        subTaskName: "walk through park",
        status: false
    }
    render(<Provider store={store}><SubTask mainTaskId={subTask.mainTaskId} index={subTask.mainTaskId} subtaskName={subTask.subTaskName} status={false} /></Provider>);
    const subTaskElement = screen.getByTestId('subTask-1');
    expect(subTaskElement).toBeInTheDocument();
    expect(subTaskElement).toHaveTextContent("walk through park");
});

cleanup();

test('snapshot matched', ()=>{
    const mainTask = {
        taskName: "walk dog",
        id: 1,
        status: false,
        subtasks: [
            {
                subTaskName: "walk through park",
                status: false
            },
            {
                subTaskName: "get back home",
                status: false
            }
        ],
    }
    const tree = TestRenderer.create(<Provider store={store}><MainTask showSubtasks={mainTask.showSubtasks} taskName={mainTask.taskName} id={mainTask.id} status={mainTask.status} subtasks={mainTask.subtasks} /></Provider>).toJSON
    expect(tree).toMatchSnapshot();
});

cleanup();
