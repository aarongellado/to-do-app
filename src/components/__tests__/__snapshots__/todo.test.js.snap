// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot matched 1`] = `[Function]`;
